static/js/cohort_details.js

```diff
@@ -1093,47 +1105,157 @@ require([
         return false;
     });
 
-    //
+    var ANONYMOUS_FILTERS = {};
+
+    var save_anonymous_filters = function()
+    {
+        // Collect all selected filters and save to session storage
+        var filters = [];
+        $('.selected-filters .panel-body span.filter-token').each(function() {
+            var $this = $(this);
+            var value = {
+                'feature': { name: $this.data('feature-name'), id: $this.data('feature-id') },
+                'value'  : { name: $this.data('value-name')  , id: $this.data('value-id')   },
+                'program': { name: $this.data('prog-name')   , id: $this.data('prog-id')    }
+            };
+            filters.push(value);
+        });
+
+        var filterStr = JSON.stringify(filters);
+        sessionStorage.setItem('anonymous_filters', filterStr);
+    };
+
+    var load_tabs_queue = [];
+    var load_anonymous_filters = function()
+    {
+        // Load anonymous filters from session storage and clear it, so it is not always there
+        var str = sessionStorage.getItem('anonymous_filters');
+        ANONYMOUS_FILTERS = JSON.parse(str);
+        sessionStorage.removeItem('anonymous_filters');
+
+        // Find out the tabs need to load to support anonymous filters
+        if (ANONYMOUS_FILTERS != null && ANONYMOUS_FILTERS.length > 0) {
+            for (i = 0; i < ANONYMOUS_FILTERS.length; ++i) {
+                var program_id = ANONYMOUS_FILTERS[i].program.id;
+                if (!load_tabs_queue.includes(program_id))
+                {
+                    load_tabs_queue.push(program_id);
+                }
+            }
+        }
+    };
+
+    var apply_anonymous_filters = function(active_program_id)
+    {
+        // Check if anonymous filter exist, then find all checkbox and check them
+        if (ANONYMOUS_FILTERS != null && ANONYMOUS_FILTERS.length > 0) {
+            for (i = 0; i < ANONYMOUS_FILTERS.length; ++i) {
+                var aFilter = ANONYMOUS_FILTERS[i];
+                if (aFilter.program.id != active_program_id)
+                    continue;
+
+                var programId = aFilter.program.id.toString();
+                var featureId = aFilter.feature.id.toString();
+                var valueId = aFilter.value.id.toString();
+
+                var checkboxId = programId + "-" + featureId + "-" + valueId;
+                checkboxId = checkboxId.replace(/ /g, "_");
+                checkboxId = checkboxId.toUpperCase();
+
+                // Get checkbox by id, jQuery does not work for some special characters so using DOM here
+                var checkbox = document.getElementById(checkboxId);
+
+                if (checkbox != null) {
+                    // Set checked and trigger change to update other related data
+                    checkbox.checked = true;
+                    var event = new Event('change');
+                    checkbox.dispatchEvent(event);
```

Comment:
The issue here is this triggers the update callback once for each filter value - a serious issue if a user clicked multiple filters. A better way to do this is to track which filter you're on in the set, and on the last one, pass a special value to trigger the update only then. Doing this requires a couple small changes:

- Add a parameter to the `filter_change_callback` event handler, eg. withoutDisplayUpdates. No default value needed.
- At the end of `filter_change_callback`, add `!withoutDisplayUpdates &&` in front of `update_displays()`
- In your calls here, you do something like: `$(checkboxId).trigger('change', [Boolean(i !== (ANONYMOUS_FILTERS.length-1))])` (Example assumes the checkbox was already set to checked and the name constructed as a selector ala `#2-VITAL_STATUS-ALIVE`)

Reply:
Good Point.  I have changed the code. Thanks!